lib/AST/ASTPrinter.cpp

```diff
@@ -4043,15 +4069,489 @@ void PrintAST::visitMissingMemberDecl(MissingMemberDecl *decl) {
   Printer << " */";
 }
 
+void PrintAST::visitIntegerLiteralExpr(IntegerLiteralExpr *expr) {
+  Printer << expr->getDigitsText();
+}
+
+void PrintAST::visitFloatLiteralExpr(FloatLiteralExpr *expr) {
+  Printer << expr->getDigitsText();
+}
+
+void PrintAST::visitNilLiteralExpr(NilLiteralExpr *expr) {
+  Printer << "nil";
+}
+
+void PrintAST::visitStringLiteralExpr(StringLiteralExpr *expr) {
+  Printer << "\"" << expr->getValue() << "\"";
+}
+
+void PrintAST::visitBooleanLiteralExpr(BooleanLiteralExpr *expr) {
+  if (expr->getValue()) {
+    Printer << "true";
+  } else {
+    Printer << "false";
+  }
+}
+
+void PrintAST::visitRegexLiteralExpr(RegexLiteralExpr *expr) {
+  Printer << expr->getRegexText();
+}
+
+void PrintAST::visitErrorExpr(ErrorExpr *expr) {
+  Printer << "<error>";
+}
+
+void PrintAST::visitIfExpr(IfExpr *expr) {
+}
+
+void PrintAST::visitIsExpr(IsExpr *expr) {
+}
+
+void PrintAST::visitTapExpr(TapExpr *expr) {
+}
+
+void PrintAST::visitTryExpr(TryExpr *expr) {
+  Printer << "try ";
+  visit(expr->getSubExpr());
+}
+
+void PrintAST::visitCallExpr(CallExpr *expr) {
+  visit(expr->getFn());
+  Printer << "(";
+  auto argList = expr->getArgs();
+  bool isFirst = true;
+  // FIXME: handle trailing closures.
+  if (argList) {
+    for (auto i = argList->begin(), iEnd = argList->end(); i != iEnd; ++i) {
+      auto arg = *i;
```

Comment:
Note you can also iterate it as `for (auto arg : *argList)`. You could even write this as:

```cxx
llvm::interleave(*argList, [&](auto arg) { printArgument(arg); },
                 [&]() { Printer << ", "; });
```

to automatically handle the intermixing of the commas. You could also use `llvm::interleave` in a couple of the below methods, though I don't feel too strongly about it.

Reply:
Updated to use the simpler iteration. Not using `interleave` for now, since existing code in this file doesn't either. I can update it to use `interleave` if people feel strongly about that, of course.